### PR TITLE
When buffer closes, focus on parent buffer

### PR DESCRIPTION
**Pull request by wes-adams**
Issue: https://github.com/helix-editor/helix/issues/4539

~~this solution adds a new field to the `Tree::Node` struct: `spawned_by`. when a new view is created, the new view's `spawned_by` field is populated with the `ViewId` of the view that was in focus at the time of creation. when a view is closed, i.e. `:q`, focus will move to the `spawned_by` view, provided it still exists. if it no longer exists, `Tree::focus_next()` is called, which moves focus to an arbitrary view that exists (as best i can tell).~~

2nd attempt:
~~i've implemented `focus_prev`. when a view closes, and it is the view in focus, `focus_prev` is called. it attempts to shift focus to the "`(current_view - 1)`" view (if it exists. if it does not, the legacy method `focus_next` is called).~~

3rd attempt:
see @the-mikedavis solution below. thanks for the help

### CODE REVIEW

**Comment by archseer** [comment]
I'd just focus the split to the left/whatever split is in the same container as the view. If it's the only view then the next one in the parent. We shouldn't need to track this on the view itself

**Comment by archseer** [comment]
Or just implementing `focus_prev` would probably be enough

**Comment by wes-adams** [comment]
> Or just implementing `focus_prev` would probably be enough

thanks for the feedback. i'll rework this with this in mind.

**Comment by wes-adams** [comment]
@archseer - i'd really appreciate another look when you can

File: helix-view/src/tree.rs [hunk 1]
```diff
@@ -521,6 +521,24 @@ impl Tree {
...
+                    None => {
+                        let (key, _) = self.traverse().next().unwrap();
+                        self.focus = key;
+                    }
+                }
+            } else {
+                self.focus_next();
+            }
+        }
+    }
```
Comment (the-mikedavis) [inline helix-view/src/tree.rs:540]: You could implement DoubleEndedIterator for Traverse and then this function would be almost the same as `focus_next`

Comment (the-mikedavis) [reply helix-view/src/tree.rs:540]: i.e. it should be possible to write `focus_prev` like this:

```rust
let mut views = self
    .traverse()
    .rev()
    .skip_while(|&(id, _view)| id != self.focus)
    .skip(1);
if let Some((id, _)) = views.next() {
    self.focus = id;
} else {
    let (key, _) = self.traverse().rev().next().unwrap();
    self.focus = key;
}
```

File: helix-view/src/tree.rs [hunk 2]
```diff
@@ -521,6 +521,26 @@ impl Tree {
...
+
+        let mut views = self
+            .traverse()
+            .rev()
+            .skip_while(|&(id, _view)| id != self.focus)
+            .skip(1); // Skip focused value
+        if let Some((id, _)) = views.next() {
+            id
+        } else {
+            // extremely crude, take the first item again
```
Comment (the-mikedavis) [inline helix-view/src/tree.rs:538]:
```suggestion
            // extremely crude, take the last item
```

**Review by the-mikedavis — APPROVED** [review]

**Comment by wes-adams** [comment]
@the-mikedavis thanks for the guidance and patience. it's much appreciated

**Pull request merged**